File: web/src/lib/api/models/app/agent/Agent.ts
// Dashboard AI agent types, mirroring internal/app/aiagent and models/agent.go.

export interface AgentSession {
    id: string;
    org_id: string;
    user_id: string;
    title: string;
    context: {
        page?: string;
        resource?: string;
        model?: string;
        pending?: PendingAgentTool | null;
    };
    created_at: string;
    updated_at: string;
}

export interface PendingAgentTool {
    tool_call_id: string;
    tool_name: string;
    risk: string;
    args_summary?: string;
}

export interface AgentSessionsPage {
    data: AgentSession[];
    pagination: { next_cursor: string | null; has_more: boolean };
}

// A persisted transcript block, hydrated server-side into the same shape the
// live stream renders (see internal/app/aiagent HydratedBlock).
export interface AgentHydratedBlock {
    kind: "text" | "tool";
    text?: string;
    tool?: string;
    args_summary?: string;
    result?: string;
    entity_type?: string;
    entity_id?: string;
    open_url?: string;
    done: boolean;
}

export interface AgentHydratedTurn {
    role: "user" | "assistant";
    blocks: AgentHydratedBlock[];
}

// GET /ai/sessions/:id/messages — a reopened session's history + any pending
// approval, so a tab rehydrates identically to a live run.
export interface AgentTranscript {
    title: string;
    turns: AgentHydratedTurn[];
    pending: PendingAgentTool | null;
    free_model?: boolean;
}

// AgentStreamEvent is one SSE step from a message/approval run.
export interface AgentStreamEvent {
    type:
        | "text"
        | "tool_start"
        | "tool_result"
        | "approval_required"
        | "iteration"
        | "error"
        | "done";
    text?: string;
    tool?: string;
    risk?: string;
    args_summary?: string;
    tool_call_id?: string;
    result?: string;
    iteration?: number;
    credits_remaining?: number;
    budget?: number;
    // True when the run is on a free/local model (AI_LOCAL_MODEL): the client
    // warns the user and no credits are charged.
    free_model?: boolean;
    code?: string;
    message?: string;
    entity_type?: string;
    entity_id?: string;
    open_url?: string;
}

File: internal/models/agent.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// AgentSession is one dashboard-agent conversation. Sessions are per-user
// (private to the member who started them); the org scope is for tenancy only.
type AgentSession struct {
	ID        uuid.UUID           `json:"id"`
	OrgID     uuid.UUID           `json:"org_id"`
	UserID    uuid.UUID           `json:"user_id"`
	Title     string              `json:"title"`
	Context   AgentSessionContext `json:"context"`
	CreatedAt time.Time           `json:"created_at"`
	UpdatedAt time.Time           `json:"updated_at"`
}

// AgentSessionContext is the read-then-execute jsonb blob on a session: the
// client's page/resource awareness, the model chosen for the run, and any tool
// call paused awaiting approval. Validated at the app boundary (this struct),
// never filtered in SQL.
type AgentSessionContext struct {
	// Page / Resource mirror the presence shape the client already pushes
	// ({page, resource}); injected into the system prompt as context.
	Page     string `json:"page,omitempty"`
	Resource string `json:"resource,omitempty"`
	// Model is the provider model id resolved for this session's tier.
	Model string `json:"model,omitempty"`
	// FreeModel is true when the session ran on a free/local backend
	// (AI_FREE): persisted so a reopened tab still shows the warning.
	FreeModel bool `json:"free_model,omitempty"`
	// Pending is the tool call awaiting the user's approve/deny when a run is
	// paused; nil when the session is idle or running.
	Pending *PendingAgentTool `json:"pending,omitempty"`
}

// PendingAgentTool is a write/send tool call paused for human approval.
type PendingAgentTool struct {
	MessageID   string          `json:"message_id"`
	ToolCallID  string          `json:"tool_call_id"`
	ToolName    string          `json:"tool_name"`
	Risk        string          `json:"risk"`
	Args        json.RawMessage `json:"args"`
	ArgsSummary string          `json:"args_summary,omitempty"`
}

// AgentMessageRow is one persisted transcript turn. Content is the serialized
// provider-agnostic message (role, text, tool_calls, tool result) so a run
// resumes losslessly after an approval pause.
type AgentMessageRow struct {
	ID        uuid.UUID       `json:"id"`
	SessionID uuid.UUID       `json:"session_id"`
	Role      string          `json:"role"`
	Content   json.RawMessage `json:"content"`
	Tokens    int             `json:"tokens"`
	CreatedAt time.Time       `json:"created_at"`
}

// AIToolPolicy is a per-org "always allow this tool" decision. Only write-class
// tools can have a policy; send-class tools are never auto-allowed.
type AIToolPolicy struct {
	OrgID     uuid.UUID  `json:"org_id"`
	ToolName  string     `json:"tool_name"`
	Decision  string     `json:"decision"`
	CreatedBy *uuid.UUID `json:"created_by,omitempty"`
	CreatedAt time.Time  `json:"created_at"`
}
